File: backend/pipeline/feeds.py
# -*- coding: utf-8 -*-
"""资讯源（Feed）的单一真源：一张 FEEDS 表 + 四个共用骨架函数。

术语（见仓库根 CONTEXT.md）：
  - **资讯源（Feed）**：定期抓取、按日份累积历史的内容来源（AI 日报 / 每日60秒 / HN / …）。
  - **日份（day）**：history 里的一条，形如 {date, fetchedAt, count, items|sections, …}。

此前这个概念在两层各被抄了七遍：`export_data.py` 七个 `get_<源>()`（213 行里约 190 行是同一条
「合并日份、留 14 天」的重复），七个 `fetch_<源>.py` 的 `main()`（同一段 try/except/写盘/打印）。
本模块把**不变的骨架**收进函数：读源 → 空壳兜底 → upsert 当天日份 → 倒序留 14 天；抓取失败不
覆盖旧文件、成功原子写。**随源变的部分**留在两处：FEEDS 表的字段（键名、文案、空壳默认值、
预览方式）与各 `fetch_<源>.py` 的 `build()`（怎么取数、怎么映射成统一 item）。

加一个新资讯源 = 写一个 `build()` + 在 FEEDS 加一行 + 在 `backend/core/paths.py` 加一个路径常量。

**为什么写盘只留一条通道**：此前七个 fetcher 各自裸 `json.dump`，而 `enrich.py` 用
`write_json_atomic` 回写同一批文件——同一文件两条写通道（一原子一非原子），读者可能读到写了
一半的 json。`run_fetcher` 统一走原子写，这条 bug 面就此消失。
"""
import json
import os
import subprocess
from dataclasses import dataclass
from typing import Callable, Tuple

from backend.core import config as wb_config
from backend.utils import common as wb_common

HISTORY_DAYS = 14  # history 保留多少个日份


def items_preview(width=50, tip=False):
    """返回一个「打印前 5 条标题」的预览函数；tip=True 额外打印「一言」（每日60秒专用）。"""
    def _preview(data):
        for i, it in enumerate(data["items"][:5], 1):
            print("   %d. %s" % (i, it["title"][:width]))
        if tip and data["tip"]:
            print("   一言：%s" % data["tip"][:40])
    return _preview


def sections_preview(data):
    """按节打印（分节结构专用，如 AI 日报）。"""
    for s in data["sections"]:
        print("   - %s (%d)" % (s["label"], len(s["items"])))


@dataclass(frozen=True)
class FeedSpec:
    """一个资讯源随源而变的那些事实。骨架行为在本模块的函数里，不在这里。"""

    key: str  # data.json 里的键，也是 FEEDS 的索引
    label: str  # 中文名，用于控制台 [OK]/[WARN] 文案
    canonical: str  # 源站链接（fetcher 与空壳共用，避免两处各写一遍）
    path_attr: str  # backend/core/paths.py 里对应的路径常量名
    source: str = ""  # 来源串；"" 表示该源的空壳与日份都没有这个键（AI 日报即如此）
    kind: str = "items"  # items | sections（AI 日报按节分组）
    hist_extra: Tuple[str, ...] = ()  # 除骨架外还要透传进日份的键（每日60秒的 tip）
    shell_extra: Tuple[str, ...] = ()  # 空壳独有的额外空串键（每日60秒的 tip/cover）
    preview: Callable = None  # 成功后的控制台预览；None = items_preview(50)

    def shell(self):
        """源文件缺失/损坏时的空壳——键与键序和此前七处手写版逐字一致。"""
        d = {"date": "", "fetchedAt": "", "count": 0, self.kind: []}
        if self.source:
            d["source"] = self.source
        d["canonical"] = self.canonical
        for k in self.shell_extra:
            d[k] = ""
        return d

    def day(self, d):
        """从源文件内容取出一个日份（history 里的一条）。"""
        out = {"date": d.get("date"), "fetchedAt": d.get("fetchedAt"),
               "count": d.get("count"), self.kind: d.get(self.kind)}
        for k in self.hist_extra:
            out[k] = d.get(k, "")
        if self.source:
            out["source"] = d.get("source", "")
        out["canonical"] = d.get("canonical", "")
        return out

    def cjk_label(self):
        """英文名与中文之间补一个空格，中文名不补（保持既有文案逐字不变）。"""
        return self.label + " " if self.label[-1].isascii() else self.label


FEEDS = {s.key: s for s in (
    FeedSpec(key="aiDaily", label="AI 日报", path_attr="AI_DAILY_JSON",
             canonical="https://aihot.virxact.com/daily",
             kind="sections", preview=sections_preview),
    FeedSpec(key="dailyNews", label="每日新闻", path_attr="DAILY_NEWS_JSON",
             canonical="https://github.com/vikiboss/60s",
             source="每日60秒 (vikiboss/60s)",
             hist_extra=("tip",), shell_extra=("tip", "cover"),
             preview=items_preview(40, tip=True)),
    FeedSpec(key="hackerNews", label="Hacker News", path_attr="HACKER_NEWS_JSON",
             canonical="https://news.ycombinator.com/",
             source="Hacker News (via OpenCLI)"),
    FeedSpec(key="githubTrending", label="GitHub Trending", path_attr="GITHUB_TRENDING_JSON",
             canonical="https://github.com/trending",
             source="GitHub Trending (via OpenCLI)"),
    FeedSpec(key="productHunt", label="Product Hunt", path_attr="PRODUCTHUNT_JSON",
             canonical="https://www.producthunt.com",
             source="Product Hunt (Atom feed)"),
    FeedSpec(key="sspai", label="少数派", path_attr="SSPAI_JSON",
             canonical="https://sspai.com",
             source="少数派 (sspai.com RSS)", preview=items_preview(40)),
    FeedSpec(key="x", label="X/推特", path_attr="X_JSON",
             canonical="https://x.com/",
             source="X (via grok-cli)"),
)}


def merge_history(src_path, data_json, spec):
    """读源文件 → 合并日份（upsert 当天、倒序、留 HISTORY_DAYS 天）→ 返回该源在 data.json 里的片段。

    history 随 data.json 经 sync.py 用 GitHub API 推送天然持久化（无需额外文件/本地 git），
    所以每次 export 都要从「上一次生成的 data.json」把历史读回来。
    源文件缺失/损坏 → 空壳，不抛异常、不影响其余资讯源（优雅劣化）。
    """
    try:
        d = json.load(open(src_path, encoding="utf-8"))
    except Exception:
        d = spec.shell()
    hist = []
    if os.path.isfile(data_json):
        try:
            hist = (json.load(open(data_json, encoding="utf-8")).get(spec.key) or {}).get("history", [])
        except Exception:
            hist = []
    if d.get("date"):
        hist = [h for h in hist if h.get("date") != d["date"]]
        hist.append(spec.day(d))
        hist.sort(key=lambda x: x.get("date", ""), reverse=True)
        hist = hist[:HISTORY_DAYS]
    d["history"] = hist
    return d


def run_fetcher(spec, build, out):
    """跑一个资讯源的抓取：失败不覆盖旧文件（返回 1），成功原子写（返回 0）。

    失败即优雅劣化——该源沿用上一次结果，export_data 与其余资讯源不受影响。
    """
    try:
        data = build()
    except Exception as e:
        print("[WARN] %s抓取失败，保留上一次结果：%s" % (spec.cjk_label(), e))
        if os.path.isfile(out):
            print("       已有 %s，未覆盖" % out)
        return 1
    wb_common.write_json_atomic(out, data)
    if spec.kind == "sections":
        print("[OK] %s %s · %d 条 · %d 节 -> %s"
              % (spec.label, data["date"], data["count"], len(data["sections"]), out))
    else:
        print("[OK] %s %s · %d 条 -> %s" % (spec.label, data["date"], data["count"], out))
    (spec.preview or items_preview())(data)
    return 0


def opencli_rows(site, command, *args, timeout=60):
    """经 OpenCLI 取一个站点的行对象数组；未配置/失败抛异常（由 run_fetcher 兜底跳过）。

    OpenCLI 需 Node>=20，仅在**取数层**被调用、不进 App 运行时（守北极星，见 ADR 0007）。
    退出码（sysexits）：0 成功 / 66 空结果 / 69 Bridge 未起 / 77 需登录；非 0 一律当异常。
    """
    cmd = wb_config.opencli_cmd()
    if not cmd:
        raise RuntimeError(
            "未配置 OpenCLI（env WB_OPENCLI_CMD / workbench.local.json opencliCmd / "
            "PATH 上的 opencli 均缺失）——本机若未装 Node>=20 + OpenCLI，该源自动跳过")
    argv = [*cmd, site, command, *args, "-f", "json"]
    try:
        p = subprocess.run(
            argv, capture_output=True, text=True,
            encoding="utf-8", errors="replace", timeout=timeout,
        )
    except Exception as e:
        raise RuntimeError("调用 OpenCLI 失败：%s" % e)
    if p.returncode == 66:
        raise ValueError("OpenCLI 返回空结果（exit 66）")
    if p.returncode != 0:
        tail = ((p.stderr or "").strip().splitlines() or [""])[-1]
        raise RuntimeError("OpenCLI 退出码 %d：%s" % (p.returncode, tail))
    try:
        rows = json.loads(p.stdout)
    except Exception as e:
        raise ValueError("OpenCLI 输出非 JSON：%s" % e)
    if not isinstance(rows, list) or not rows:
        raise ValueError("OpenCLI 输出不是非空数组")
    return rows

File: js/core/feeds.js
// 资讯源（Feed）表：前端这一侧的单一真源。术语定义见仓库根 CONTEXT.md。
//
// 后端对应表在 backend/pipeline/feeds.py（键名 / 空壳 / 日份合并），两处按 key 对齐。
// 此前 info.js 把这张表拆成七组近同构的 render/缓存/DateChanged/日期回落（约 400 行），
// infosearch.js 里还另有一张源名表；加一个新源要在前端改 6 处。
//
// 加一个新资讯源 = 这里加一行 + index.html 的 #col-info 里加一个 <div id="<id>Block">。
//
// 字段说明（都是"随源而变的事实"，渲染骨架在 js/views/info.js 的 renderFeed）：
//   key        data.json 里的键（与后端 FEEDS 的 key 一致）
//   id         DOM/localStorage 前缀：<id>Block / <id>Body / <id>Sel，折叠态也用它当键
//              —— 改它会让用户已保存的折叠状态失效，别动
//   kind       sections（按节分组，仅 AI 日报）| items
//   label      源标题条与正文大标题用的名字
//   unit       计数单位：条 / 个 / 篇
//   icon       内联 SVG 图标名（见 core/icons.js）
//   histLabel  历史下拉的前缀文案
//   fallbackSource / linkText   正文 meta 行的来源兜底名与外链文案
//   cardTitle / cardIcon        条目列表卡片的标题与图标（sections 源不用）
//   defaultSrc 条目没有自带 source 时显示的兜底来源；""=不显示（AI 日报如此）
//   searchLabel 检索面板里这个源的名字（infosearch.js 用）
//   ask        「让 AI 讲讲」的提问开头
//   summary    条目是否展示摘要（每日60秒刻意不展示）
//   numbered   条目是否编号（AI 日报按节分组，不编号）
//   empty      没数据时的提示 {text（含 HTML，是本仓字面量不转义）, cmd（可选：一键让 AI 抓）}
//   actions    顶部「提炼 / 存库」按钮组；null = 该源没有这组按钮
//              prompt 里的 {n} 会被当日条目数替换
export const FEEDS = [
  {
    key: "aiDaily", id: "news", kind: "sections",
    label: "AI 日报", unit: "条", icon: "fileText", histLabel: "历史日报",
    fallbackSource: "AI HOT", linkText: "看完整日报 ↗",
    defaultSrc: "", searchLabel: "AI 日报",
    ask: "用大白话展开讲讲这条 AI 新闻的背景和影响，并说说对我有什么用：",
    summary: true, numbered: false,
    empty: {
      text: "这一天还没有抓到日报数据。可以点「立即刷新」让本机重新抓一次；也可以让 WorkBuddy 手动跑 <code>fetch_ai_daily.py</code>。",
      cmd: "跑一下 personal-workbench 的 fetch_ai_daily.py 抓今天的 AI 日报，然后 export + push",
    },
    actions: {
      title: "基于日报做点什么", refine: "提炼 3 条要点", archive: "存进知识库",
      refinePrompt: "下面是今天的 AI 日报（{n} 条）。挑出对我最有用的 3 条要点，各配一个今天就能动手试的小实验。",
      archivePrompt: "把今天这份 AI 日报整理成一篇可归档的主题化摘要（Markdown，分主题、要点式）。整理好后我自己点「存知识库」入库。",
    },
  },
  {
    key: "dailyNews", id: "dnews", kind: "items",
    label: "每日新闻", unit: "条", icon: "fileText", histLabel: "历史新闻",
    fallbackSource: "每日60秒", linkText: "看来源 ↗",
    cardTitle: "今日头条", cardIcon: "trendingUp",
    defaultSrc: "每日60秒", searchLabel: "每日60秒",
    ask: "用大白话展开讲讲这条新闻的背景，并说说对我有什么影响：",
    summary: false, numbered: true, cover: true, tip: true,
    empty: {
      text: "这一天还没有抓到新闻数据。可以点「立即刷新」让本机重新抓一次；也可以让 WorkBuddy 手动跑 <code>fetch_daily_news.py</code>。",
      cmd: "跑一下 personal-workbench 的 fetch_daily_news.py 抓今天的国内新闻，然后 export + push",
    },
    actions: {
      title: "基于新闻做点什么", refine: "挑 3 条相关的", archive: "存进知识库",
      refinePrompt: "下面是今天的每日新闻（{n} 条）。挑 3 条跟我最相关的，说说为什么值得关注。",
      archivePrompt: "把今天这份每日新闻整理成一篇可归档的主题化摘要（Markdown）。整理好后我自己点「存知识库」入库。",
    },
  },
  {
    key: "hackerNews", id: "hnews", kind: "items",
    label: "Hacker News 热帖", unit: "条", icon: "trendingUp", histLabel: "历史热帖",
    fallbackSource: "Hacker News", linkText: "去 HN ↗",
    cardTitle: "今日热帖", cardIcon: "trendingUp",
    defaultSrc: "Hacker News", searchLabel: "Hacker News",
    ask: "用大白话讲讲这条 Hacker News 热帖在讨论什么、为什么值得关注：",
    summary: true, numbered: true, actions: null,
    empty: { text: "还没有抓到 Hacker News 数据。点「立即刷新」让本机经 OpenCLI 抓一次；若本机没装 OpenCLI（Node ≥20），该源会自动跳过，不影响其它资讯。" },
  },
  {
    key: "githubTrending", id: "gt", kind: "items",
    label: "GitHub Trending 今日热门", unit: "个", icon: "trendingUp", histLabel: "历史热门",
    fallbackSource: "GitHub Trending", linkText: "去 Trending ↗",
    cardTitle: "今日热门仓库", cardIcon: "trendingUp",
    defaultSrc: "GitHub Trending", searchLabel: "GitHub Trending",
    ask: "用大白话讲讲这个 GitHub 仓库是做什么的、解决了什么问题、什么时候该用它：",
    summary: true, numbered: true, actions: null,
    empty: { text: "还没有抓到 GitHub Trending 数据。点「立即刷新」让本机经 OpenCLI 抓一次；若本机没装 OpenCLI（Node ≥20），该源会自动跳过，不影响其它资讯。" },
  },
  {
    key: "productHunt", id: "ph", kind: "items",
    label: "Product Hunt 每日新品", unit: "个", icon: "trendingUp", histLabel: "历史新品",
    fallbackSource: "Product Hunt", linkText: "去 Product Hunt ↗",
    cardTitle: "今日上新（挑一个拆产品感）", cardIcon: "trendingUp",
    defaultSrc: "Product Hunt", searchLabel: "Product Hunt",
    ask: "用大白话讲讲这个新产品解决了谁的什么需求、亮点在哪、我能从它的设计里学到什么：",
    summary: true, numbered: true, actions: null,
    empty: { text: "还没有抓到 Product Hunt 数据。点「立即刷新」让本机抓一次（走公开 Atom feed，无需 Node/OpenCLI）。" },
  },
  {
    key: "sspai", id: "sspai", kind: "items",
    label: "少数派上新", unit: "篇", icon: "fileText", histLabel: "历史上新",
    fallbackSource: "少数派", linkText: "去少数派 ↗",
    cardTitle: "近期文章", cardIcon: "fileText",
    defaultSrc: "少数派", searchLabel: "少数派",
    ask: "用大白话讲讲这篇文章在说什么、对提升产品品味/效率有什么启发：",
    summary: true, numbered: true, actions: null,
    empty: { text: "还没有抓到少数派数据。点「立即刷新」让本机抓一次（走公开 RSS，无需 Node/OpenCLI）。" },
  },
  {
    key: "x", id: "x", kind: "items",
    label: "X/推特热议", unit: "条", icon: "messageCircle", histLabel: "历史热议",
    fallbackSource: "X (via grok-cli)", linkText: "去 X ↗",
    cardTitle: "近期热议（挑一条深挖）", cardIcon: "messageCircle",
    defaultSrc: "X", searchLabel: "X",
    ask: "用大白话讲讲这条 X 帖子在说什么、背景是什么、值不值得我关注：",
    summary: true, numbered: true, actions: null,
    empty: { text: "还没有抓到 X 数据。该源经 grok-cli（需 Bun + xAI 付费 key）抓取，未配置时自动跳过——在 <code>workbench.local.json</code> 填好 grokCmd / grokApiKey 后，点「立即刷新」即可。" },
  },
];

// 按 DOM 前缀取源（内联 onclick 只传得动字符串，回调里靠它还原 spec）
export function feedById(id) {
  return FEEDS.filter(function (s) { return s.id === id; })[0] || null;
}
